### ISSUE TITLE
Prometheus Crash Looping on startup

### ISSUE THREAD

**Issue by ridersofrohan** [GoogleCloudPlatform/prometheus-engine#459]
<img width="1645" alt="Screen Shot 2023-04-17 at 11 16 03" src="https://user-images.githubusercontent.com/3170854/232531592-8cc6194a-c655-47f6-a4aa-cbba94137efe.png">

* Related to https://github.com/prometheus/prometheus/issues/6934

I'm trying to change the resource limits but since it is managed, I do not have the permissions to edit / modify anything in the `gke-gmp-system` namespace.

How can I reset the pod so that it wipes the wal directory. There was an issue where it was scraping too many timeseries, and caused it to get into this crash looping situation. I limited the number of timeseries but now the collector is stuck and can't progress.

Any recommendations?

**Comment by bwplotka** [comment]
Hey, 

Unless I am wrong here - it seems the collector does NOT have any persistent storage at the moment (see https://github.com/GoogleCloudPlatform/prometheus-engine/blob/main/manifests/operator.yaml#L657), so I am surprised there IS anything to wipe out (and causing crash loop),  unless you are using something custom here. Are you sure you are interpreting the root cause of crash loop correctly? Maybe the high cardinality scrape is still happening? What is the size of the node (unfortunately we have static limit at the moment, so for extremely big nodes our limit might not enough).

If there would be persistent storage (host dir or PV) the solutions would be:
* Remove PV
* Increase limit (which you tried)
*  Wipe the WAL manually. To do so you need to attach debug container to the collector and delete [`/prometheus/data/wal`] directory (or whole `prometheus/data`)

However, as I mentioned, something else might be happening as there is no persistent storage for collector as far as I know (: Hope that helps!

**Comment by ego93** [comment]
I have a similar issue where I see collector pods crash looping on startup but then become stable.
Looking at the logs seems to suggest the `config-reloader` health check fails before prometheus service has started, triggering a pod restart.

Logs:
```
level=info ts=2023-07-20T06:19:59.058365954Z caller=main.go:82 msg="ensure ready-url is healthy
level=error ts=2023-07-20T06:19:59.560166348Z caller=main.go:91 msg="polling ready-url" err="Get \"http://localhost:19090/-/ready\": dial tcp 127.0.0.1:19090: connect: connection refused
...
ts=2023-07-20T06:21:49.020Z caller=main.go:966 level=info msg="Server is ready to receive web requests.
```

**Comment by pintohutch** [comment]
Hi @ego93 - yes this is a known issue with a tentative fix in https://github.com/GoogleCloudPlatform/prometheus-engine/pull/474

**Comment by bwplotka** [comment]
The OOM due to WAL replay/potential WAL corruption was fixed in main (https://github.com/GoogleCloudPlatform/prometheus-engine/pull/947) and being released as 0.13.0 version. Will close this issue after initial version will be available on latest GKE (or manual install).

Note that it's still possible to trigger OOM simply by adding excessive amount of targets/metrics to scrape to a single agent. The same version will have optional VPA you can configure to avoid this if you have such cases. 

Feedback and reports welcome around OOM problems, but in separate issues please 🤗 Thanks!

**Comment by bwplotka** [comment]
Finally this has been rolled as part of the [0.13.0](https://github.com/GoogleCloudPlatform/prometheus-engine/releases/tag/v0.13.0#:~:text=%5BENHANCEMENT%5D%20Collectors%20now%20skip%20WAL%20playback%20on%20startup%2C%20which%20can%20avoid%20slow%20startup%20and%20crashloop%20issues.) release.

Essentially Prometheus should not escape the OOM crashloop IF you reduce the number of targets you are consuming. Even if you don't it will not immediately OOM AND allow for less often OOMs, as there is no replay anymore (which sometimes consumed more memory than normal usage).

Please open new issuses for subsequent OOMs related issues for managed collection GMP 0.13+ Thanks!
